### ISSUE TITLE
One method for operator evaluation

### ISSUE THREAD

**Issue by kohr-h** [odlgroup/odl#57]
Based on the discussion in #48:

I'll make a branch to unify `_apply()` and `_call()` in `Operator` as as one `_call()` method:

- one positional argument (`x`)  --> in-place
- two positional arguments (`x, out`)  --> out-of-place
- two positional arguments, second optional (`x, out=None`)  --> either way, depending on `out`

`kwargs` would be passed through from `__call__()`.

**Comment by adler-j** [comment]
Yes, do this on a branch and we'll evaluate it.

**Comment by kohr-h** [comment]
Do we want to allow both extra `*args` and `**kwargs` or just `**kwargs`?

**Comment by adler-j** [comment]
Id say only kwargs, args can cause issues with out.

**Comment by kohr-h** [comment]
:+1: Saves my ass :wink:

**Comment by kohr-h** [comment]
I finally [got it working](https://github.com/odlgroup/odl/tree/issue-57__operator), still with some debugging code and prints, and of course, operator tests will fail now. 

Most of the work (in time and code lines) was to write the [dispatcher for the argument spec](https://github.com/odlgroup/odl/blob/issue-57__operator/odl/operator/operator.py#L85-L253) for `_call()`, in particular for Python 3, where you can make arguments keyword-only. This raises the number of possibilities significantly, but at least I don't need to inspect variable args, phew!

So the new list of possibilities is 

1. one positional argument `(x)`  --> in-place
2. two positional arguments `(x, out)`  --> out-of-place
3. two positional arguments, second optional `(x, out=None)`  --> either way, depending on `out`
4. one positional argument and `out` as kw-only `(x, *, out=None)`  --> same as 3.
5. two kw-only arguments `(*, x=None, out=None)` --> same as 3. May not make so much sense to have a default for `x`, but anyway... Yeah, maybe this case can be excluded. Would simplify the logic.

Rewriting the `Operator` class was not so hard in the end. The [logic deciding how to set up the instance](https://github.com/odlgroup/odl/blob/issue-57__operator/odl/operator/operator.py#L338-L356) now happens in the `__new__()` method of `Operator`, where the methods `_call_in_place` and `_call_out_of_place` are set according to how `_call` can be used. I tested all the variants, and it seems to do what it should. More tests to come, but not today.

Some more thought had to go into the `operator()` function (I renamed it to `simple_operator`), but there the solution was actually quite easy: use the same dispatcher for the given `call` function (TODO: remove default value `None` for that argument) and write three simple classes which implement nothing but `_call()`, with signature depending on the case. See [here](https://github.com/odlgroup/odl/blob/issue-57__operator/odl/operator/operator.py#L1611-L1633). 

Not sure if this is better or worse than using a metaclass, but one advantage is that Pylint has a chance to understand the code.

And, from a user's point of view, I guess it's easier to implement one function instead of two.

**Comment by kohr-h** [comment]
Actually, when I think again, both Py3-specific patterns (4. and 5.) don't make much sense as there is no way to enforce keyword-only args. Hold on, I'll throw them out.

**Comment by kohr-h** [comment]
Okay, we can keep 4. to make it easier to use such functions e.g. in `simple_operator`.

Example (with debugging output) - out-of-place (oop):

```python
>>> from builtins import super
>>> class OpOut(odl.Operator):
...     def __init__(self, mat):
...         self.mat = mat
...         dom = odl.Rn(mat.shape[1])
...         ran = odl.Rn(mat.shape[0])
...         super().__init__(dom, ran)
...         
...     def _call(self, x):
...         return self.range.element(self.mat.dot(x))

>>> mat = [1, 2, 3] * np.eye(3)

>>> op_out = OpOut(mat)
ArgSpec(args=['self', 'x'], varargs=None, keywords=None, defaults=None)
has_out:  False
out_optional:  False

>>> x = odl.Rn(3).one()
>>> op_out(x)
in op.__call__: out-of-place
Rn(3).element([1.0, 2.0, 3.0])

>>> out = odl.Rn(3).element()
>>> op_out(x, out)
in op.__call__: in-place
use _default_call_in_place
Rn(3).element([1.0, 2.0, 3.0])
```

Example in-place:

```python
>>> class OpIn(odl.Operator):
...     def __init__(self, mat):
...         self.mat = mat
...         dom = odl.Rn(mat.shape[1])
...         ran = odl.Rn(mat.shape[0])
...         super().__init__(dom, ran)
...         
...     def _call(self, x, out):
...         self.mat.dot(x, out.asarray())

>>> mat = [1, 2, 3] * np.eye(3)

>>> op_in = OpIn(mat)
ArgSpec(args=['self', 'x', 'out'], varargs=None, keywords=None, defaults=None)
has_out:  True
out_optional:  False

>>> op_in(x)
in op.__call__: out-of-place
use _default_call_out_of_place
Rn(3).element([1.0, 2.0, 3.0])

>>> op_in(x, out)
in op.__call__: in-place
Rn(3).element([1.0, 2.0, 3.0])
```

Example dual-use:

```python
>>> class OpBoth(odl.Operator):
...     def __init__(self, mat):
...         self.mat = mat
...         dom = odl.Rn(mat.shape[1])
...         ran = odl.Rn(mat.shape[0])
...         super().__init__(dom, ran)
...         
...     def _call(self, x, out=None):
...         if out is None:
...             return self.range.element(self.mat.dot(x))
...         else:
...             self.mat.dot(x, out.asarray())
...             return out

>>> mat = [1, 2, 3] * np.eye(3)

>>> op_both = OpBoth(mat)
ArgSpec(args=['self', 'x', 'out'], varargs=None, keywords=None, defaults=(None,))
has_out:  True
out_optional:  True

>>> op_both(x)
in op.__call__: out-of-place
Rn(3).element([1.0, 2.0, 3.0])

>>> op_both(x, out)
in op.__call__: in-place
Rn(3).element([1.0, 2.0, 3.0])
```

In Python 3:

```python
>>> class OpPy3(odl.Operator):
...     def __init__(self, mat):
...         self.mat = mat
...         dom = odl.Rn(mat.shape[1])
...         ran = odl.Rn(mat.shape[0])
...         super().__init__(dom, ran)
...     def _call(self, x, *, out=None):
...         if out is None:
...             return self.range.element(self.mat.dot(x))
...         else:
...             self.mat.dot(x, out=out.asarray())
...             return out

>>> mat = [1, 2, 3] * np.eye(3)
>>> op_py3 = OpPy3(mat)
FullArgSpec(args=['self', 'x'], varargs=None, varkw=None, defaults=None, kwonlyargs=['out'], kwonlydefaults={'out': None}, annotations={})
has_out:  True
out_optional:  True

>>> op_py3(x)
in op.__call__: out-of-place
Rn(3).element([1.0, 2.0, 3.0])

>>> op_py3(x, out)
in op.__call__: in-place
Rn(3).element([1.0, 2.0, 3.0])
```

**Comment by adler-j** [comment]
Also closes like 10 other issues :D

**Comment by kohr-h** [comment]
Right, need to go through that list.

**Comment by kohr-h** [comment]
Closed by #116
